src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java

```diff
@@ -0,0 +1,464 @@
+ * if the some of the required caches are out of date, or a non-caching computable node is encountered.
+ * The latter performs the required evaluations along the way.
+ *
+ * IMPORTANT NOTE: the queried values are return by reference. It is the user's responsibility not to mutate
+ * them. Otherwise, the functional structure will be broken.
+ *
+ * Mutation:
+ * =========
+ *
+ * The primitive values can be mutated by calling {@link ImmutableComputableGraph#setValue(String, Duplicable)}.
+ * The mutations do not occur in place; rather, a new instance of {@link ImmutableComputableGraph} is created along
+ * with new instances for the updated nodes. If cache auto update is enabled, the affected nodes will be evaluated.
+ * Otherwise, only the cache status will go out of date. Unchanged nodes are passed as reference to the new instance.
+ * JVM's garbage collector will free up the memory for old cached nodes in sequential computations.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public class ImmutableComputableGraph implements Serializable {
+
+    private static final long serialVersionUID = -1162776031416105027L;
+
+    private static Map<String, Duplicable> EMPTY_MAP = new HashMap<>();
+
+    /**
+     * A simple builder class for {@link ImmutableComputableGraph}
+     */
+    public static class ImmutableComputableGraphBuilder {
+        private final Collection<CacheNode> nodes;
+        private boolean cacheAutoUpdate;
+
+        ImmutableComputableGraphBuilder() {
+            nodes = new ArrayList<>();
+            cacheAutoUpdate = false;
+        }
+
+        public ImmutableComputableGraphBuilder addPrimitiveNode(@Nonnull final String key,
+                                                                @Nonnull final String[] tags,
+                                                                @Nonnull Duplicable value) {
+            nodes.add(new PrimitiveCacheNode(key, Arrays.stream(tags).collect(Collectors.toList()), value));
+            return this;
+        }
+
+        public ImmutableComputableGraphBuilder addComputableNode(@Nonnull final String key,
+                                                                 @Nonnull final String[] tags,
+                                                                 @Nonnull final String[] parents,
+                                                                 @Nullable final Function<Map<String, ? extends Duplicable>, ? extends Duplicable> func,
+                                                                 final boolean cacheEvals) {
+            nodes.add(new ComputableCacheNode(key,
+                    Arrays.stream(tags).collect(Collectors.toList()),
+                    Arrays.stream(parents).collect(Collectors.toList()),
+                    func, cacheEvals));
+            return this;
+        }
+
+        public ImmutableComputableGraphBuilder enableCacheAutoUpdate() {
```

Comment:
Yes, but they are part of the ICG functionality that we may want to return for future possible use / prototyping.